clients/pkg/logentry/stages/stage.go

```diff
@@ -107,134 +109,119 @@ func toStage(p Processor) Stage {
 func New(logger log.Logger, jobName *string, stageType string,
 	cfg interface{}, registerer prometheus.Registerer) (Stage, error) {
-	var s Stage
-	var err error
-	switch stageType {
-	case StageTypeDocker:
-		s, err = NewDocker(logger, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeCRI:
-		s, err = NewCRI(logger, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeJSON:
-		s, err = newJSONStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeLogfmt:
-		s, err = newLogfmtStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeRegex:
-		s, err = newRegexStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeMetric:
-		s, err = newMetricStage(logger, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeLabel:
-		s, err = newLabelStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeLabelDrop:
-		s, err = newLabelDropStage(cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeTimestamp:
-		s, err = newTimestampStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeOutput:
-		s, err = newOutputStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeMatch:
-		s, err = newMatcherStage(logger, jobName, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeTemplate:
-		s, err = newTemplateStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeTenant:
-		s, err = newTenantStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeReplace:
-		s, err = newReplaceStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeDrop:
-		s, err = newDropStage(logger, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeSampling:
-		s, err = newSamplingStage(logger, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeLimit:
-		s, err = newLimitStage(logger, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeMultiline:
-		s, err = newMultilineStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypePack:
-		s, err = newPackStage(logger, cfg, registerer)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeLabelAllow:
-		s, err = newLabelAllowStage(cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeStaticLabels:
-		s, err = newStaticLabelsStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeDecolorize:
-		s, err = newDecolorizeStage(cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeEventLogMessage:
-		s, err = newEventLogMessageStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	case StageTypeGeoIP:
-		s, err = newGeoIPStage(logger, cfg)
-		if err != nil {
-			return nil, err
-		}
-	default:
+	creator, ok := getStageCreators()[stageType]
```

Comment:
nit: IIUC getStageCreatorsn is a factory. I think it's more idiomatic if it takes the stageType as an argument and returns the stage right away instead fo a map.

Reply:
Moreover I think this function should take the stageType and the params and return the created stage.

Reply:
refactored in https://github.com/grafana/loki/pull/9986/commits/3dd5be56156c2a788a4b2153c29783aa650e5f74